src/components/tx/SignOrExecuteForm/hooks.ts

```diff
@@ -0,0 +1,88 @@
+import { useMemo } from 'react'
+import { type TransactionOptions, type SafeTransaction } from '@safe-global/safe-core-sdk-types'
+import useSafeInfo from '@/hooks/useSafeInfo'
+import useWallet from '@/hooks/wallets/useWallet'
+import useOnboard from '@/hooks/wallets/useOnboard'
+import { isSmartContractWallet } from '@/hooks/wallets/wallets'
+import {
+  dispatchOnChainSigning,
+  dispatchTxExecution,
+  dispatchTxProposal,
+  dispatchTxSigning,
+} from '@/services/tx/tx-sender'
+import { useHasPendingTxs } from '@/hooks/usePendingTxs'
+import { sameString } from '@safe-global/safe-core-sdk/dist/src/utils'
+
+type TxActions = {
+  signTx: (safeTx: SafeTransaction, txId?: string) => Promise<string>
+  executeTx: (safeTx: SafeTransaction, txId: string | undefined, txOptions: TransactionOptions) => Promise<string>
+}
+
+export const useTxActions = (): TxActions => {
+  const { safe, safeAddress } = useSafeInfo()
+  const { chainId, version } = safe
+  const onboard = useOnboard()
+  const wallet = useWallet()
+
+  return useMemo<TxActions>(() => {
+    const proposeTx = async (sender: string, safeTx: SafeTransaction, txId?: string, origin?: string) => {
+      const tx = await dispatchTxProposal({
+        chainId,
+        safeAddress,
+        sender,
+        safeTx,
+        txId,
+        origin,
+      })
+      return tx.txId
+    }
+
+    const signTx: TxActions['signTx'] = async (safeTx, txId) => {
+      if (!onboard) throw 'Onboard not ready'
+      if (!wallet) throw 'Wallet not connected'
+
+      // Smart contract wallets must sign via an on-chain tx
+      if (await isSmartContractWallet(wallet)) {
+        // If the first signature is a smart contract wallet, we have to propose w/o signatures
+        // Otherwise the backend won't pick up the tx
+        // The signature will be added once the on-chain signature is indexed
+        const id = txId || (await proposeTx(wallet.address, safeTx, txId, origin))
+        await dispatchOnChainSigning(safeTx, id, onboard, chainId)
+        return id
+      }
+
+      // Otherwise, sign off-chain
+      const signedTx = await dispatchTxSigning(safeTx, version, onboard, chainId, txId)
+      return await proposeTx(wallet.address, signedTx, txId, origin)
+    }
+
+    const executeTx: TxActions['executeTx'] = async (safeTx, txId, txOptions) => {
+      if (!onboard) throw 'Onboard not ready'
+      if (!wallet) throw 'Wallet not connected'
```

Comment:
I would suggest to throw with an `Error` object to stay consistent with the othe throws.